File: app/src/main/java/com/italankin/dictionary/utils/SharedPrefs.java
package com.italankin.dictionary.utils;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.italankin.dictionary.api.ApiClient;
import com.italankin.dictionary.dto.Language;

import java.io.File;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.lang.reflect.Type;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import io.reactivex.Completable;
import io.reactivex.Single;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

/**
 * Wrapper around {@link SharedPreferences} for this application purposes.
 */
public class SharedPrefs {

    private static final String LANGS_FILE_NAME = "langs.json";
    private static final SimpleDateFormat LANGS_TIMESTAMP =
            new SimpleDateFormat("yyyy-MM-dd'T'HH:MM:ssZ", Locale.US);

    private static final String PREF_SOURCE = "source";
    private static final String PREF_DEST = "dest";
    private static final String PREF_LANGS_LOCALE = "langs_locale";
    private static final String PREF_LANGS_TIMESTAMP = "langs_timestamp";
    private static final String PREF_LOOKUP_REVERSE = "lookup_reverse";
    private static final String PREF_BACK_FOCUS = "back_focus";
    private static final String PREF_CLOSE_ON_SHARE = "close_on_share";
    private static final String PREF_INCLUDE_TRANSCRIPTION = "include_transcription";
    private static final String PREF_FILTER_FAMILY = "filter_family";
    private static final String PREF_FILTER_SHORT_POS = "filter_short_pos";
    private static final String PREF_FILTER_MORPHO = "filter_morpho";
    private static final String PREF_FILTER_POS_FILTER = "filter_pos_filter";
    private static final String PREF_SHOW_SHARE_FAB = "show_share_fab";

    private final File filesDir;
    private final SharedPreferences preferences;
    private final Gson gson = new Gson();

    private Disposable mSaveLanguagesSub;

    public SharedPrefs(Context context) {
        filesDir = context.getFilesDir();
        preferences = PreferenceManager.getDefaultSharedPreferences(context);
    }

    ///////////////////////////////////////////////////////////////////////////
    // Destination language
    ///////////////////////////////////////////////////////////////////////////

    public void setDestLang(String code) {
        SharedPreferences.Editor editor = preferences.edit();
        editor.putString(PREF_DEST, code);
        editor.apply();
    }

    public void setDestLang(Language lang) {
        setDestLang(lang.getCode());
    }

    public String getDestLang() {
        return preferences.getString(PREF_DEST, null);
    }

    ///////////////////////////////////////////////////////////////////////////
    // Source language
    ///////////////////////////////////////////////////////////////////////////

    public void setSourceLang(String code) {
        SharedPreferences.Editor editor = preferences.edit();
        editor.putString(PREF_SOURCE, code);
        editor.apply();
    }

    public void setSourceLang(Language lang) {
        setSourceLang(lang.getCode());
    }

    public String getSourceLang() {
        return preferences.getString(PREF_SOURCE, null);
    }

    ///////////////////////////////////////////////////////////////////////////
    // Languages list
    ///////////////////////////////////////////////////////////////////////////

    /**
     * Save list of languages on disk.
     *
     * @param list list of languages
     */
    public void saveLanguagesList(List<Language> list) {
        if (list == null) {
            return;
        }
        if (mSaveLanguagesSub == null || mSaveLanguagesSub.isDisposed()) {
            mSaveLanguagesSub = Completable
                    .fromCallable(() -> {
                        Collections.sort(list);
                        File file = getLangsFile();
                        if (!file.delete()) {
                            Timber.d("saveLanguagesList: delete failed");
                        }
                        FileOutputStream fs = new FileOutputStream(file);
                        String json = gson.toJson(list);
                        fs.write(json.getBytes());
                        fs.close();
                        String locale = Locale.getDefault().getLanguage();
                        preferences.edit().putString(PREF_LANGS_LOCALE, locale).apply();
                        return true;
                    })
                    .subscribeOn(Schedulers.io())
                    .subscribe(() -> Timber.d("saveLanguagesList: success"),
                            throwable -> Timber.e(throwable, "saveLanguagesList:"));
        }
    }

    public Single<List<Language>> getLanguagesList() {
        return Single
                .fromCallable(() -> {
                    File file = getLangsFile();
                    FileReader fs = new FileReader(file);
                    Type collectionType = new TypeToken<List<Language>>() {}.getType();
                    return gson.<List<Language>>fromJson(fs, collectionType);
                })
                .doOnError(throwable -> {
                    purgeLanguages();
                    Timber.e(throwable, "getLanguagesList:");
                });
    }

    public void setLangsTimestamp(Date date) {
        String timestamp = LANGS_TIMESTAMP.format(date);
        preferences.edit().putString(PREF_LANGS_TIMESTAMP, timestamp).apply();
    }

    public boolean shouldUpdateLangs() {
        boolean updatedLastTwoWeeks = false;
        if (preferences.contains(PREF_LANGS_TIMESTAMP)) {
            try {
                String s = preferences.getString(PREF_LANGS_TIMESTAMP, null);
                Date timestamp = LANGS_TIMESTAMP.parse(s);
                Calendar calendar = Calendar.getInstance();
                calendar.add(Calendar.DAY_OF_MONTH, -14);
                updatedLastTwoWeeks = timestamp.after(calendar.getTime());
            } catch (ParseException e) {
                // failed to parse date
            }
        }
        String savedLocale = preferences.getString(PREF_LANGS_LOCALE, null);
        String currentLocale = Locale.getDefault().getLanguage();
        return !updatedLastTwoWeeks || !currentLocale.equals(savedLocale) || !hasLangsFile();
    }

    public boolean hasLangsFile() {
        return getLangsFile().exists();
    }

    public void purgeLanguages() {
        if (mSaveLanguagesSub != null && !mSaveLanguagesSub.isDisposed()) {
            mSaveLanguagesSub.dispose();
        }
        File file = getLangsFile();
        if (file.exists()) {
            Timber.d("purgeLanguages: %s", file.delete());
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    // Other
    ///////////////////////////////////////////////////////////////////////////

    public boolean showShareFab() {
        return preferences.getBoolean(PREF_SHOW_SHARE_FAB, true);
    }

    public boolean lookupReverse() {
        return preferences.getBoolean(PREF_LOOKUP_REVERSE, true);
    }

    public boolean backFocusSearch() {
        return preferences.getBoolean(PREF_BACK_FOCUS, false);
    }

    public boolean closeOnShare() {
        return preferences.getBoolean(PREF_CLOSE_ON_SHARE, false);
    }

    public boolean shareIncludeTranscription() {
        return preferences.getBoolean(PREF_INCLUDE_TRANSCRIPTION, false);
    }

    public int getSearchFilter() {
        int family = preferences.getBoolean(PREF_FILTER_FAMILY, true) ? ApiClient.FILTER_FAMILY : 0;
        int shortPos = preferences.getBoolean(PREF_FILTER_SHORT_POS, false) ? ApiClient.FILTER_SHORT_POS : 0;
        int morpho = preferences.getBoolean(PREF_FILTER_MORPHO, false) ? ApiClient.FILTER_MORPHO : 0;
        int pos = preferences.getBoolean(PREF_FILTER_POS_FILTER, false) ? ApiClient.FILTER_POS_FILTER : 0;
        return family | shortPos | morpho | pos;
    }

    ///////////////////////////////////////////////////////////////////////////
    // Internal
    ///////////////////////////////////////////////////////////////////////////

    private File getLangsFile() {
        return new File(filesDir, LANGS_FILE_NAME);
    }

}

File: app/src/main/java/com/italankin/dictionary/utils/SharedPrefs.kt
package com.italankin.dictionary.utils

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.italankin.dictionary.api.FILTER_FAMILY
import com.italankin.dictionary.api.FILTER_MORPHO
import com.italankin.dictionary.api.FILTER_POS_FILTER
import com.italankin.dictionary.api.FILTER_SHORT_POS
import com.italankin.dictionary.api.dto.Language
import io.reactivex.Completable
import io.reactivex.Single
import io.reactivex.disposables.Disposable
import io.reactivex.rxkotlin.subscribeBy
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import java.io.File
import java.io.FileOutputStream
import java.io.FileReader
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

/**
 * Wrapper around [SharedPreferences] for this application purposes.
 */
class SharedPrefs(
        context: Context,
        private val gson: Gson
) {

    companion object {
        private const val LANGS_FILE_NAME = "langs.json"
        private val LANGS_TIMESTAMP = SimpleDateFormat("yyyy-MM-dd'T'HH:MM:ssZ", Locale.US)

        private const val PREF_SOURCE = "source"
        private const val PREF_DEST = "dest"
        private const val PREF_LANGS_LOCALE = "langs_locale"
        private const val PREF_LANGS_TIMESTAMP = "langs_timestamp"
        private const val PREF_LOOKUP_REVERSE = "lookup_reverse"
        private const val PREF_BACK_FOCUS = "back_focus"
        private const val PREF_CLOSE_ON_SHARE = "close_on_share"
        private const val PREF_INCLUDE_TRANSCRIPTION = "include_transcription"
        private const val PREF_FILTER_FAMILY = "filter_family"
        private const val PREF_FILTER_SHORT_POS = "filter_short_pos"
        private const val PREF_FILTER_MORPHO = "filter_morpho"
        private const val PREF_FILTER_POS_FILTER = "filter_pos_filter"
        private const val PREF_SHOW_SHARE_FAB = "show_share_fab"
    }

    private val filesDir: File = context.filesDir
    private val preferences: SharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)

    private var saveLanguagesDisposable: Disposable? = null

    fun setDestLang(lang: Language) {
        destLang = lang.code
    }

    var destLang: String?
        get() = preferences.getString(PREF_DEST, null)
        set(code) = preferences.edit().putString(PREF_DEST, code).apply()

    fun setSourceLang(lang: Language) {
        sourceLang = lang.code
    }

    var sourceLang: String?
        get() = preferences.getString(PREF_SOURCE, null)
        set(code) = preferences.edit().putString(PREF_SOURCE, code).apply()

    val languagesList: Single<List<Language>>
        get() = Single
                .fromCallable {
                    val file = getLanguagesFile()
                    val fs = FileReader(file)
                    val collectionType = object : TypeToken<List<Language>>() {}.type
                    gson.fromJson<List<Language>>(fs, collectionType)
                }
                .doOnError {
                    Timber.e(it, "getLanguagesList:")
                    purgeLanguages()
                }

    fun setLanguagesTimestamp(date: Date?) {
        val timestamp = LANGS_TIMESTAMP.format(date)
        preferences.edit().putString(PREF_LANGS_TIMESTAMP, timestamp).apply()
    }

    val shouldUpdateLanguages: Boolean
        get() {
            var updatedLastTwoWeeks = false
            if (preferences.contains(PREF_LANGS_TIMESTAMP)) {
                try {
                    val timestamp = preferences.getString(PREF_LANGS_TIMESTAMP, null)
                    val timestampDate = LANGS_TIMESTAMP.parse(timestamp)
                    val calendar = Calendar.getInstance()
                    calendar.add(Calendar.DAY_OF_MONTH, -14)
                    updatedLastTwoWeeks = timestampDate.after(calendar.time)
                } catch (e: ParseException) { // failed to parse date
                    Timber.e(e, "shouldUpdateLangs:")
                }
            }
            val savedLocale = preferences.getString(PREF_LANGS_LOCALE, null)
            val currentLocale = Locale.getDefault().language
            return !updatedLastTwoWeeks || currentLocale != savedLocale || !getLanguagesFile().exists()
        }

    fun saveLanguagesList(list: List<Language>?) {
        if (list == null) {
            return
        }
        val disposable = saveLanguagesDisposable
        if (disposable == null || disposable.isDisposed) {
            saveLanguagesDisposable = Completable
                    .fromRunnable {
                        val file = getLanguagesFile()
                        FileOutputStream(file).use {
                            val sorted = list.sorted()
                            val json = gson.toJson(sorted)
                            it.write(json.toByteArray())
                        }
                        val locale = Locale.getDefault().language
                        preferences.edit().putString(PREF_LANGS_LOCALE, locale).apply()
                    }
                    .subscribeOn(Schedulers.io())
                    .subscribeBy(
                            onComplete = { Timber.d("saveLanguagesList: success") },
                            onError = { throwable -> Timber.e(throwable, "saveLanguagesList:") }
                    )
        }
    }

    val showShareFab: Boolean
        get() = preferences.getBoolean(PREF_SHOW_SHARE_FAB, true)

    val lookupReverse: Boolean
        get() = preferences.getBoolean(PREF_LOOKUP_REVERSE, true)

    val backFocusSearch: Boolean
        get() = preferences.getBoolean(PREF_BACK_FOCUS, false)

    val closeOnShare: Boolean
        get() = preferences.getBoolean(PREF_CLOSE_ON_SHARE, false)

    val shareIncludeTranscription: Boolean
        get() = preferences.getBoolean(PREF_INCLUDE_TRANSCRIPTION, false)

    val searchFilter: Int
        get() {
            val family = if (preferences.getBoolean(PREF_FILTER_FAMILY, true)) FILTER_FAMILY else 0
            val shortPos = if (preferences.getBoolean(PREF_FILTER_SHORT_POS, false)) FILTER_SHORT_POS else 0
            val morpho = if (preferences.getBoolean(PREF_FILTER_MORPHO, false)) FILTER_MORPHO else 0
            val pos = if (preferences.getBoolean(PREF_FILTER_POS_FILTER, false)) FILTER_POS_FILTER else 0
            return family or shortPos or morpho or pos
        }

    private fun getLanguagesFile(): File = File(filesDir, LANGS_FILE_NAME)

    private fun purgeLanguages() {
        if (saveLanguagesDisposable != null && !saveLanguagesDisposable!!.isDisposed) {
            saveLanguagesDisposable!!.dispose()
        }
        val file = getLanguagesFile()
        if (file.exists()) {
            Timber.d("purgeLanguages: %s", file.delete())
        }
    }
}
